### ISSUE TITLE
Could not find a compatible version of MSBuild.

### ISSUE THREAD

**Issue by NelzkieCoder** [xenko3d/xenko#47]
Can't start xenko. Kept having this error

Could not find a compatible version of MSBuild.

**Comment by LuizZak** [comment]
I ran into this issue earlier.

I found instructions to solve a [similar issue described here](https://developercommunity.visualstudio.com/content/problem/94979/cannot-open-any-projects-in-solution-after-install.html):

> 1) open Developer Command Prompt for VS2017 (I opened it as admin)
> 2) gacutil /u Microsoft.Build.Framework
> to uninstall the Microsoft.Build.Framework from Global Assembly Cache [GAC]
>
> at this point I checked with "gacutil /l Microsoft.Build.Framework" that the v15.1 of that assembly was removed. I also opened VS2017, and the projects in my solution opened properly. However I wanted the latest& greatest Microsoft.Build.Framework back, so I did
>
> 3) gacutil /i "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Microsoft.Build.Framework.dll"
> 
> to re-install the 15.1.0 version.

It seems to work, now.

**Comment by jhorv** [comment]
I was having this and decided to try updating all the Microsoft.Build.* packages.  That worked for me so I submitted [a PR for it](https://github.com/xenko3d/xenko/pull/52).

**Comment by jazzay** [comment]
I am getting this error even after syncing master with the above fixes to package versions. 

![image](https://user-images.githubusercontent.com/554363/43754516-236eb6dc-99c0-11e8-937e-bf665a0c8d3b.png)

I have VS Community 15.7.6 installed.

**Comment by jazzay** [comment]
My MS Build version is newer than that referenced in the project files:
**Microsoft (R) Build Engine version 15.7.180.61344 for .NET Framework**

**Comment by xen2** [comment]
@jazzay you can even try latest `master` which should display more info about why it failed in the messagebox

**Comment by koenigvonamburu** [comment]
Here is the exception message that i get with the latest master. The "System.Runtime.InteropServices.RuntimeInformation.dll" is located in the same folder as the Xenko.GameStudio.exe executable and the version 4.0.1.0 is fitting. Maybe one of it´s dependecies is not available? The TypeInitializationException does not tell which one...

![image](https://user-images.githubusercontent.com/6361533/43795700-754e45ea-9a82-11e8-9154-86f6e34a7811.png)

**Comment by xen2** [comment]
To know the actual .NET version (which could be different from Windows file version), you have to open it with a tool such as [ILSpy](https://github.com/icsharpcode/ILSpy/releases)

**Comment by jazzay** [comment]
Ok via IL Spy is shows:

![image](https://user-images.githubusercontent.com/554363/43816370-35712512-9a89-11e8-86e3-5ea61d95f3c3.png)

**Comment by xen2** [comment]
What I have on my computer:

**Xenko folder:**
```
// C:\Users\xen\.nuget\packages\xenko\3.0.0.1\Bin\Windows\System.Runtime.InteropServices.RuntimeInformation.dll
// System.Runtime.InteropServices.RuntimeInformation, Version=4.0.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
```

**Xenko.GameStudio.exe.config**
```
        <assemblyIdentity name="System.Runtime.InteropServices.RuntimeInformation" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.2.0" />
```

**MSBuild folder:**
```
// C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\System.Runtime.InteropServices.RuntimeInformation.dll
// System.Runtime.InteropServices.RuntimeInformation, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
```

**Comment by xen2** [comment]
@jazzay which one did you open, MSBuild one or Xenko one?
If you have a few minutes, please stop by [gitter](https://gitter.im/xenko3d/xenko) so that we can try to solve it together.

**Comment by xen2** [comment]
Yes however during his build it seems to have chosen 4.0.2.0:
`Encountered conflict between 'CopyLocal:C:\Program Files\dotnet\sdk\NuGetFallbackFolder\system.runtime.interopservices.runtimeinformation\4.3.0\lib\net45\System.Runtime.InteropServices.RuntimeInformation.dll' and 'CopyLocal:C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\net462\lib\System.Runtime.InteropServices.RuntimeInformation.dll'.  Choosing 'CopyLocal:C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\Microsoft\Microsoft.NET.Build.Extensions\net462\lib\System.Runtime.InteropServices.RuntimeInformation.dll' because AssemblyVersion '4.0.2.0' is greater than '4.0.1.0'.`

**Comment by xen2** [comment]
I remember we had some similar issues related to the fact we output everything in same folder.
Also various conflicts with the nuget fallback and some VS overrides for some packages.

**Comment by jazzay** [comment]
Note I was able to solve my problem by installing Xenko 3.0 from the website (as recommended by @xen2 ) I had not installed Xenko on my machine prior.

**Comment by Kryptos-FR** [comment]
> related to the fact we output everything in same folder

That's another item to add to #12.

I remember one of the reason is that some dependencies are/were not explicit between some projects/deps and that without it some DLLs would be missing in the runtime or editor. But that does create mismatch with assembly versions depending on the order of build. There might be a better way.

**Comment by NelzkieCoder** [comment]
Has this been fixed? I am still having the error

**Comment by xen2** [comment]
@Kryptos-FR The actual issue is that some project uses .NET Standard libraries and some others are not. Using .NET Standard triggers lot of various libraries being swapped internally (with different versions).
And yes, #12 would switch to a model where each library output in its own folder -- following how project work by default will make it easier to maintain. However we'll need to override assembly resolution but it should be easier to maintain (and necessary anyway if we want to support plugins or user-built assemblies replacement).

**Comment by xen2** [comment]
@NelzkieCoder Did you try installing and running official Xenko to see if prerequisites help?

**Comment by NelzkieCoder** [comment]
Did a reset of my PC. And install VS2017 with latest update. Everything works. However when I run a template. No screen is showing.

**Comment by koenigvonamburu** [comment]
@PeaceSells50 goto  sources/editor/Xenko.GameStudio/app.config and replace ` <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.2.0" />`  with ` <bindingRedirect oldVersion="0.0.0.0-4.0.2.0" newVersion="4.0.1.0" />` (maybe you can also remove the redirect). Rebuild and run.

**Comment by Kryptos-FR** [comment]
@koenigvonamburu that is not going to work. The Game Studio needs version 4.0.2.0 to properly function.

There is an issue with a test project that copies the wrong version of the DLL. @xen2 is already looking into it.

In the meantime to workaround, make sure to not compile any test projects. Do a clean first and always compile the `Xenko.GameStudio` project (and eventually the tools projects if they are not built).

In any case if you encounter the issue again, go to the bin folder where Xenko is built, delete the `System.Runtime.InteropServices.RuntimeInformation.dll` there as well as its copy in the Direct3D11 subfolder and recompile `Xenko.GameStudio`.

**Comment by xen2** [comment]
This should be fixed in latest Xenko 3.0.0.7 and 3.1 beta (or a previous version).
Please reopen if you still have the issue after an update.

**Comment by pogsdbest** [comment]
Still Encounter The Error when opening Xenko 3.1.0.1 and  3.0

Xenko 3.1.0.1

Could not find a compatible version of MSBuild.
Check that you have a valid installation with the required workloads, or go to www.visualstudio.com/downloads to install a new one.
System.IO.FileNotFoundException: Could not load file or assembly 'System.Numerics.Vectors, Version=4.1.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified. File name: 'System.Numerics.Vectors, Version=4.1.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' at System.SpanHelpers.IndexOf(Char& searchSpace, Char value, Int32 length) at Microsoft.Build.Evaluation.Expander2.Function1.ExtractPropertyFunction(String expressionFunction, IElementLocation elementLocation, Object propertyValue, UsedUninitializedProperties usedUnInitializedProperties, IFileSystem fileSystem) at Microsoft.Build.Evaluation.Expander2.PropertyExpander1.ExpandPropertyBody(String propertyBody, Object propertyValue, IPropertyProvider1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties, IFileSystem fileSystem)
   at Microsoft.Build.Evaluation.Expander2.PropertyExpander1.ExpandPropertiesLeaveTypedAndEscaped(String expression, IPropertyProvider1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties, IFileSystem fileSystem) at Microsoft.Build.Evaluation.Expander2.PropertyExpander1.ExpandPropertiesLeaveEscaped(String expression, IPropertyProvider1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties, IFileSystem fileSystem)
   at Microsoft.Build.Evaluation.Expander2.ExpandIntoStringLeaveEscaped(String expression, ExpanderOptions options, IElementLocation elementLocation) at Microsoft.Build.Evaluation.ToolsetReader.ExpandPropertyUnescaped(ToolsetPropertyDefinition property, Expander2 expander)
   at Microsoft.Build.Evaluation.ToolsetReader.EvaluateAndSetProperty(ToolsetPropertyDefinition property, PropertyDictionary1 properties, PropertyDictionary1 globalProperties, PropertyDictionary1 initialProperties, Boolean accumulateProperties, String& toolsPath, String& binPath, Expander2& expander)
   at Microsoft.Build.Evaluation.ToolsetReader.ReadToolset(ToolsetPropertyDefinition toolsVersion, PropertyDictionary1 globalProperties, PropertyDictionary1 initialProperties, Boolean accumulateProperties)
   at Microsoft.Build.Evaluation.ToolsetReader.ReadEachToolset(Dictionary2 toolsets, PropertyDictionary1 globalProperties, PropertyDictionary1 initialProperties, Boolean accumulateProperties) at Microsoft.Build.Evaluation.ToolsetReader.ReadToolsets(Dictionary2 toolsets, PropertyDictionary1 globalProperties, PropertyDictionary1 initialProperties, Boolean accumulateProperties, String& msBuildOverrideTasksPath, String& defaultOverrideToolsVersion)
   at Microsoft.Build.Evaluation.ToolsetReader.ReadAllToolsets(Dictionary2 toolsets, ToolsetRegistryReader registryReader, ToolsetConfigurationReader configurationReader, PropertyDictionary1 environmentProperties, PropertyDictionary1 globalProperties, ToolsetDefinitionLocations locations) at Microsoft.Build.Evaluation.ProjectCollection.InitializeToolsetCollection(ToolsetRegistryReader registryReader, ToolsetConfigurationReader configReader) at Microsoft.Build.Evaluation.ProjectCollection..ctor(IDictionary2 globalProperties, IEnumerable1 loggers, IEnumerable`1 remoteLoggers, ToolsetDefinitionLocations toolsetDefinitionLocations, Int32 maxNodeCount, Boolean onlyLogCriticalEvents, Boolean loadProjectsReadOnly) at Microsoft.Build.Evaluation.ProjectCollection..ctor() at Xenko.Core.Assets.PackageSessionPublicHelper.CheckMSBuildToolset() at Xenko.GameStudio.Program.<Startup>d__12.MoveNext() in D:\BuildAgent\work\8cafb933a54b01d0\sources\editor\Xenko.GameStudio\Program.cs:line 241
WRN: Assembly binding logging is turned OFF. To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1. Note: There is some performance penalty associated with assembly bind failure logging. To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].
